README.md
<img src="https://github.com/AraBlocks/docs/blob/master/ara.png" width="30" height="30" /> ara-network-node-dns
====================

[![Build Status](https://travis-ci.com/AraBlocks/ara-network-node-dns.svg?token=r6p7pesHZ9MRJsVsrYFe&branch=master)](https://travis-ci.com/AraBlocks/ara-network-node-dns)

An Ara Network node that runs a DNS.

## Status
This project is still in alpha development.

> **Important**: While this project is under active development, run `npm link` in `ara-network-node-dns` directory & `npm link ara-network-node-dns` in `ara-network` directory.

## Dependencies
- [Node](https://nodejs.org/en/download/)
- [ara-network][ara-network]

## Installation
```sh
$ npm install ara-network ara-network-node-dns
```

## Configuration
[ara-runtime-configuration][ara-runtime-configuration] is a dependency of [ara-network][ara-network] and will either read from the nearest `.ararc`.  Install [ara-runtime-configuration][ara-runtime-configuration] separately to specify default values not present in an `.ararc`.

Runtime configuration can be specified by targeting the
`[network.node.dns]` _INI_ section or a nested _JSON_ object
`{ network: { node: { dns: { ... }}}}`.

### Examples
_INI_ format in an `.ararc`:
```ini
[network.node.dns]
multicast = true
loopback = true
port = 5300
```
_JSON_ format (passed in to ara-runtime-configuration `rc` function:
```json
"network": {
  "node": {
    "dns": {
      "multicast": true,
      "loopback": true,
      "port": 5300
    }
  }
}
```

## Usage
```js
const dns = require('ara-network-node-dns')
```

#### Command Line (ann)
```bash
$ ann --type dns --help
usage: ann -t dns [options]

Options:
  --debug, -D  Enable debug output                                     [boolean]
  --conf, -C   Path to configuration file                               [string]
  --help, -h   Show help                                               [boolean]
  --port, -p   Port or ports to listen on          [number] [default: [53,5300]]
  --loopback   Loopback DNS discovery                  [boolean] [default: true]
  --multicast  Multicast DNS discovery                 [boolean] [default: true]
```

### Examples
#### Command Line (ann)
Invoke a network node with the `ann` (or `ara-network-node`) command line interface:
```sh
$ ann --type dns --port 5300
```

## Contributing
- [Commit message format](/.github/COMMIT_FORMAT.md)
- [Commit message examples](/.github/COMMIT_FORMAT_EXAMPLES.md)
- [How to contribute](/.github/CONTRIBUTING.md)

## See Also
- [ara-network][ara-network]
- [ara-runtime-configuration][ara-runtime-configuration]
- [dns-discovery](https://github.com/mafintosh/dns-discovery)

## License
LGPL-3.0

[ara-network]: https://github.com/arablocks/ara-network
[ara-runtime-configuration]: https://github.com/arablocks/ara-runtime-configuration


CHANGELOG.md
# [0.5.0](https://github.com/AraBlocks/ara-network-node-dns/compare/0.4.2...0.5.0) (2018-12-13)



## [0.4.2](https://github.com/AraBlocks/ara-network-node-dns/compare/0.4.1...0.4.2) (2018-12-12)


### Bug Fixes

* **index.js:** Clean up queries on success ([785caf8](https://github.com/AraBlocks/ara-network-node-dns/commit/785caf8))



## [0.4.1](https://github.com/AraBlocks/ara-network-node-dns/compare/0.4.0...0.4.1) (2018-12-12)


### Bug Fixes

* **index.js:** Clean up pending queries ([58cfb25](https://github.com/AraBlocks/ara-network-node-dns/commit/58cfb25))



# [0.4.0](https://github.com/AraBlocks/ara-network-node-dns/compare/0.3.1...0.4.0) (2018-12-07)


### Features

* **index.js:** Polyfill DNS proxy to known servers ([6816035](https://github.com/AraBlocks/ara-network-node-dns/commit/6816035))



## [0.3.1](https://github.com/AraBlocks/ara-network-node-dns/compare/0.3.0...0.3.1) (2018-10-19)



# [0.3.0](https://github.com/AraBlocks/ara-network-node-dns/compare/0.2.2...0.3.0) (2018-08-16)



## [0.2.2](https://github.com/AraBlocks/ara-network-node-dns/compare/0.2.1...0.2.2) (2018-08-16)



## [0.2.1](https://github.com/AraBlocks/ara-network-node-dns/compare/9645c82...0.2.1) (2018-08-16)


### Bug Fixes

* Fix `ports` to `port` typo ([3e4c200](https://github.com/AraBlocks/ara-network-node-dns/commit/3e4c200))


### Features

* **Dockerfile:** Adding Dockerfile ([a2359e2](https://github.com/AraBlocks/ara-network-node-dns/commit/a2359e2))
* **index.js:** adding option to pass port in cli ([9645c82](https://github.com/AraBlocks/ara-network-node-dns/commit/9645c82))
* **index.js:** adding option to pass port,loopback & multicast in cli ([ebc89bf](https://github.com/AraBlocks/ara-network-node-dns/commit/ebc89bf))
* **Makefile:** Adding Makefile for docker builds ([f27d015](https://github.com/AraBlocks/ara-network-node-dns/commit/f27d015))





index.js
/* eslint-disable no-param-reassign */
const { info, warn } = require('ara-console')
const debug = require('debug')('ara:network:node:dht')
const dns = require('ara-network/dns')
const DNS = require('dns-socket')
const ip = require('ip')

/**
 * Configuration of the DNS server
 */
const conf = {
  multicast: true,
  loopback: true,
  ports: [ 53, 5300 ]
}

/**
 * DNS server object
 */
let server = null

/**
 * Starts a DNS server node
 * @public
 * @param {Object} argv
 * @return Boolean
 */
async function start() {
  if (server) { return false }

  const sock = DNS()
  const ports = []

  server = dns.createServer(conf)
  server.on('error', onerror)
  server.listen(conf.port)

  for (const s of server._sockets) {
    s.on('listening', () => {
      const { port } = s.address()
      info('dns: Listening on port %s', port)
      ports.push(port)
    })
  }

  if (process.env.USE_ARAX_DNS) {
    server._onquery = function (query, port, host, socket) {
      const reply = { questions: query.questions, answers: [] }

      for (let i = 0; i < query.questions.length; i++) {
        this._onquestion(query.questions[i], port, host, reply.answers)
      }

      for (let i = 0; i < query.answers.length; i++) {
        this._onanswer(query.answers[i], port, host, socket)
      }

      for (let i = 0; i < query.additionals.length; i++) {
        this._onanswer(query.additionals[i], port, host, socket)
      }

      const servers = server.servers.slice()
      const pending = []

      kick()

      function cleanup() {
        while (pending.length) {
          sock.cancel(pending.shift())
        }
      }

      function kick() {
        if (!servers.length) {
          socket.response(query, reply, port, host)
          cleanup()
          return
        }

        const addr = servers.shift()
        const { questions, additionals } = query

        if ('localhost' === addr.host) {
          kick()
        } else {
          enqueue(addr.secondaryPort, addr.host)
          enqueue(addr.port, addr.host)
        }

        function enqueue(_port, _host) {
          if (
            'localhost' !== _host &&
            !ip.isLoopback(_host) &&
            ports.includes(_port)
          ) {
            const q = { questions, additionals }
            const id = sock.query(q, _port, _host, onquery)
            pending.push(id)
          } else {
            onquery(null, null)
          }
        }

        function onquery(err, res) {
          if (!err && res && res.answers && res.answers.length) {
            reply.answers = res.answers
            socket.response(query, reply, port, host)
            cleanup()
          } else {
            kick()
          }
        }
      }
    }
  }

  return true

  function onerror(err) {
    if (err && 'EACCES' === err.code) {
      const { port } = err
      conf.ports.splice(conf.ports.indexOf(port), 1)
      debug('error:', err)
    } else if (true === conf.multicast) {
      if (err && 'EADDRINUSE' === err.code) {
        debug('error:', err)
      }
    }
  }
}

/**
 * Stops a DNS server node
 * @public
 * @return Boolean
 */
async function stop() {
  if (!server) { return false }
  warn('dns: Stopping server')
  server.destroy(ondestroy)
  return true
  function ondestroy() {
    server = null
  }
}

/**
 * Configures a DNS server node
 * @public
 * @param {Object} opts
 * @return Object
 */
async function configure(opts, program) {
  if (program) {
    const { argv } = program
      .option('p', {
        type: 'number',
        alias: 'port',
        default: conf.ports,
        describe: 'Port or ports to listen on',
      })
      .option('L', {
        type: 'boolean',
        alias: 'loopback',
        default: conf.loopback,
        describe: 'Loopback DNS discovery',
      })
      .option('M', {
        type: 'boolean',
        alias: 'multicast',
        default: conf.multicast,
        describe: 'Multicast DNS discovery',
      })

    if (argv.port) { opts.port = argv.port }
    if (argv.loopback) { opts.loopback = argv.loopback }
    if (argv.multicast) { opts.multicast = argv.multicast }
  }

  conf.port = opts.port
  conf.loopback = opts.loopback
  opts.multicast = opts.multicast
}

/**
 * Returns reference to DNS server node
 * @public
 * @return Object
 */
async function getInstance() {
  return server
}

module.exports = {
  getInstance,
  configure,
  start,
  stop,
}
